File: python_version/pl0_scanner.py
#! /usr/bin/python3
# -*- coding: utf8 -*- 

"""
    python 版本的词法扫描器
    其实将问题抽象起来就是从一个字符串流中提取每个符号
    将其做上标记并输出的过程
"""

import sys
import re


# 规则声明

keywords     = r"const|var|procedure|begin|end|if|then|while|do|read|call|write|writeln|module|int"

constant     = r"[0-9]+"

identfier     = r"[A-Za-z_][A-Za-z_0-9]*"

operator     = r"\+|-|\*|\/|#|=|<|>|<>|:=|>=|<="

delimiter    = r"[,.;():]"

whitespace   = r"\s*"

separator    = r"([+\-*/#,.;():\s]|<=|>=|:=|=|<|>)"

print ("""
1:      keywords
2:      operator
3:      delimiter
4:      constant
5:      identfier
        """)

with open(sys.argv[1], mode='r') as pl0_file:
# 每次的词法分析只处理一行
    content = []
    result = []
    word_error = []
    content = pl0_file.readlines()
    for line in content:
        line = re.split(separator, line)
        line = [c for c in line if re.match('\S', c)]
        for word in line:
            if (re.match(keywords, word)):
                result.append((word, 1))
                continue
            if (re.match(operator, word)):
                result.append((word, 2))
                continue
            if (re.match(delimiter, word)):
                result.append((word, 3))
                continue
            if (re.match(constant, word)):
                result.append((word, 4))
                continue
            if (re.match(identfier, word)):
                result.append((word, 5))
                continue
            word_error.append("invalid word: %s"%(word))
            
        print(line)
    print(result)
    if len(word_error) != 0:
        print(word_error)

File: cpp_version/pl0_scanner.cpp
#include <iostream>
#include <algorithm>
#include <regex>
#include <vector>
#include <string>
#include <fstream>
#include <boost/algorithm/string.hpp>
#include <boost/algorithm/string/split.hpp>
#include <boost/algorithm/string/classification.hpp>


using std::string;
using std::stringstream;
using std::regex;
using std::regex_match;
using std::vector;
using std::for_each;
using std::cin;
using std::cout;
/*
 * c++ 版本的词法分析器
 */

const string digit        = R"([0-9])";

const string keywords     = R"(const|var|procedure|begin|end|if|then"
								"|while|do|read|call|write|writeln|module|int)";

const string constant     = R"([0-9]+)";

const string identfier    = R"([A-Za-z_][A-Za-Z_0-9]*)";

const string operator_tag = R"(\+|-|\*|\/|#|=|<|>|<>|:=|>=|<=)";

const string delimiter    = R"([,.;():])";

const string delim		= R"([ \n\t])";

const string whitespace	= R"([ \n\t]+)";

enum word_type
{
	keywords_,
	operator_tag_,
	delimiter_,
	constant_,
	identfier_,
	error
};

regex digit_regex(digit);
regex keywords_regex(keywords);
regex constant_regex(constant);
regex identfier_regex(identfier);
regex operator_tag_regex(operator_tag);
regex delimiter_regex(delimiter);
regex whitespace_regex(whitespace);



word_type get_word_type(const string& word)
{
	if (regex_match(word, keywords_regex))
		return word_type::keywords_;
	if(regex_match(word, operator_tag_regex))
		return word_type::operator_tag_;
	if(regex_match(word, delimiter_regex))
		return word_type::delimiter_;
	if(regex_match(word, constant_regex))
		return word_type::constant_;
	if(regex_match(word, identfier_regex))
		return word_type::identfier_;
	
	return word_type::error;
}
int main(int argc, char* argv[], char* env[])
{
	std::fstream filestream(argv[1], std::ios_base::in);
	const string separator = "+-*/,.;:()<>:=";
	string line;
	vector<string> words;
	vector<std::pair<string, word_type>> result;
	while (std::getline(filestream, line))
	{
		boost::split(words, line, boost::is_any_of(separator));
		std::for_each(words.begin(), words.end(), [&](const string& word){
					result.emplace_back(word, get_word_type(word));
				});
	}
	for_each(result.begin(), result.end(), [](const std::pair<string, word_type> word )
										{ printf("%s\t:\t%d", word.first.c_str(), word.second); });
	return 0;
}

